### PR TITLE
metal: DSV4 M1 Ultra prefill/decode campaign

### PR DESCRIPTION
**Pull request by auroter**
Prefill and decode optimization for DeepSeek V4 Flash 0731 (`dsv4-xxs-1` profile) on Apple Metal, M1 Ultra.

The point of this campaign: match and beat **antirez's ds4** — the optimized reference engine for this model — on the same box, with the full SlimServe/vLLM serving stack (OpenAI API, DSpark speculative decoding, TurboQuant draft KV, prefix caching). Both engines measured on the same M1 Ultra with the exact-token harness.

## SlimServe vs ds4 (same box, same model, same workloads)

| Workload | antirez ds4 | SlimServe (this PR) | vs ds4 |
| --- | ---: | ---: | ---: |
| Prefill, 2048-token chunk | 277 tok/s | **369 tok/s** | **133%** |
| Decode, 1k-in/2k-out anchor | 21.1 tok/s end-to-end (25.5 decode-only) | **31.6 tok/s** end-to-end (33.3 decode-only) | **150%** |

Campaign start was 79 tok/s prefill and 15 tok/s decode, so this is 4.7× / 2.1× over where the stack began.

## Where the speed comes from

**Decode (15 → 31.6 tok/s)**
- Per-step marshalling memos: dtype casts, weight-slice views, and rope tables are converted once per step/load instead of per layer. This alone roughly doubled decode — per-layer host overhead was starving the GPU.
- Fused mHC pre/post kernels: one dispatch per token replaces ~230 aten launches per call, including the full Sinkhorn iteration.
- Native `dsv4_router_topk`, sum-folded q2_K down projection (GEMV + weighted-sum in one kernel), split-K sparse-MLA decode, fp16-direct KV insert, single-dispatch inverse RoPE.
- Load-time q2_K SoA repack so the decode GEMVs read coalesced planes.

**Prefill (79 → 369 tok/s)**
- Tiled MoE prefill GEMM: a `qc_moe_mm_map0` work queue feeds a 64-slot dual-half iq2_xxs simdgroup-MMA tile and a q2_K SoA down twin — MoE layers stop being token-at-a-time GEMVs at prefill widths. Biggest single prefill win.
- Dense-causal MMA flash attention over pre-decoded half scratches, replacing the decode-shape candidate walk on prefill steps.
- Fused compressor front (cr=4 and cr=128) and a wide-tile transposed-store q8_0 GEMM for the large-M dense projections.
- Profile fix: `max_num_batched_tokens` 2048 → 2176 so the scheduler's draft-token reserve stops splitting a 2048-token prompt into two chunks.

## Also in this PR

- Cleanup pass: env switches 37 → 12, dead experiments removed end-to-end, one latent bug fixed (pointer-vs-string compare in the native tape).
- Merged with current `main` (A100 Q4_K fused MoE, NaN split-K fix, Muse-Glimmer Metal): the campaign's fast paths and Muse-Glimmer's coexist behind format/dtype dispatch, and the profile ships main's 262144-token sizing — which this branch's Metal indexer port actually makes reachable, and which measures slightly faster at the decode anchor.
- Six kernel oracle tests under `tests/kernels/` (runnable directly on the box, no pytest needed).
- Campaign records in `perf/` (method, per-wave evidence, raw-artifact locations).

## Validation

Zero-numeric-change discipline for the cleanup and the main merge: pinned to the benchmark config, the 8-tok, off1-2000, and 2500-token serving anchors are bit-exact (output sha256 + spec-decode counters) against the pre-cleanup build, and all six kernel oracles pass. The table above is measured on the merged tree under the shipping profile.

## Known issues (documented in-code; fixes coming as separate PRs)

- ~~A pre-existing async-output completion-event race can park the engine on a boot's first multi-chunk prefill~~ Fixed in-branch: on MPS the async-output copy and its completion event stay on the producing stream (`vllm/v1/worker/gpu/async_utils.py`, regression test `tests/kernels/test_metal_async_output.py`); the boot-ramp protocol in `perf/prefill_handoff.md` is marked superseded.
- The step-tape diagnostic (`VLLM_QC_STEP_TAPE`, default off) needs route re-validation before its tape-only mode is trusted; see `vllm/models/deepseek_v4/metal_tape.py`.

## Summary by CodeRabbit

* **New Features**
  * Expanded Apple Silicon support for DeepSeek-V4 attention, routing, compression, sparse indexing, and MoE workloads.
  * Added float16 support alongside bfloat16 across serving paths.
  * Added fused activation, normalization, quantized matrix operations, and optimized prefill/decode execution.
  * Added optional completion dumping and opt-in Metal performance profilers.
  * Added macOS memory-management improvements for GGUF loading.

* **Performance**
  * Reduced CPU synchronization and improved MPS batching, caching, transfers, asynchronous output handling, and decode/prefill throughput.

* **Bug Fixes**
  * Improved numerical consistency, deterministic routing, quantization scaling, and fallback behavior.

### CODE REVIEW

**Comment by ehartford** [comment]
phenomenal!

File: csrc/quixicore/metal/kernels/attention/mla/mla.metal [hunk 1]
```diff
@@ -762,7 +920,628 @@ kernel void mla_decode_fp8_sparse_two_cache_packed(
...
+                                    mla_ue8m0_scale(e);
+                            } else {
+                                v = float(rope[d - NOPE]);
+                            }
+                            sres[js][d] = v;
+                        }
+                    }
+                }
+            }
+            threadgroup_barrier(mem_flags::mem_threadgroup);
```
[the remark this answers is not in the document]
Comment (ehartford) [reply csrc/quixicore/metal/kernels/attention/mla/mla.metal:1339]: Addressed in 82debfed4 as a documented contract rather than a runtime check: the sole call site already routes non-conforming head counts away from this kernel (`batch >= 64 && heads % 16 == 0` in qc_metal_serving.mm), and the launcher now carries an explicit num_heads % 16 contract comment. tk_launch.h has no torch dependency, so the loud failure belongs at the op boundary, which already enforces it.

File: csrc/quixicore/metal/kernels/common/tk_launch.h [hunk 2]
```diff
@@ -2984,8 +2984,12 @@ void launch_mla_q_norm_rope(E& e, typename E::in_t q, typename E::in_t cos,
...
                             typename E::in_t norm_weight, typename E::out_t out,
                             int M, int num_heads, int nope_dim, int rope_dim,
-                            int norm_mode, float eps, int head_dim) {
-  e.pipeline(mla_q_norm_rope_kernel_name(head_dim));
+                            int norm_mode, float eps, int head_dim,
+                            bool half_input = false) {
+  // half_input selects the fp16-load variant (rounds to bf16 in-register;
+  // only instantiated for head_dim 512 — the DSV4 serving shape).
+  e.pipeline(half_input ? "mla_q_norm_rope_half_" + std::to_string(head_dim)
+                        : mla_q_norm_rope_kernel_name(head_dim));
```
[the remark this answers is not in the document]
Comment (ehartford) [reply csrc/quixicore/metal/kernels/common/tk_launch.h:2994]: Addressed in 82debfed4: the launcher comment now states the half_input => head_dim == 512 contract and where it is enforced — the sole caller passes the literal 512 and TORCH_CHECK-gates q's last dim, so no other head_dim can reach the half-input pipeline name.

File: csrc/quixicore/metal/kernels/moe/moe_mm_id/moe_mm_id.metal [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+    threadgroup_barrier(mem_flags::mem_threadgroup);
+    for (short j = sgitg; j < nr1b; j += 4) {
+      const int idj = ids[im * tokens + r1 + NR1 + j];
+      device half *Dr = D + (long)idj * N + r0;
+      threadgroup const float *C = sc + j * NR0;
+      for (short i = tiisg; i < nr0; i += 32) {
+        Dr[i] = (half)C[i];
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ehartford) [reply csrc/quixicore/metal/kernels/moe/moe_mm_id/moe_mm_id.metal:425]: Fixed in 82debfed4 host-side: ggml_moe_mm_id's output allocation is now at::zeros instead of at::empty, so slots dropped by map0 (negative router ids) surface as zeros — the same T(0) contract the vec kernels implement — for both tile kernels, including the q2_K scatter. Kernel-side zeroing would need a pass over slots no expert owns; the host fill is one blit and covers both routes.

[the remark this answers is not in the document]
Comment (ehartford) [reply csrc/quixicore/metal/kernels/moe/moe_mm_id/moe_mm_id.metal:425]: Revised in e28ea0afa after an independent review pass measured the zeros fill at ~107 MB per call (~8.6 GB of store per 2176-token prefill chunk step across layers) defending a case both Python callers make unreachable (`expert_map is None` gates both mm routes, and negative ids only arise via expert_map). The allocation returns to at::empty and the non-negative-id contract is now documented on both the host allocation and the fused_moe.py route.

File: vllm/v1/worker/gpu/model_states/default.py [hunk 4]
```diff
@@ -207,5 +221,6 @@ def prepare_attn(
             mm_req_doc_ranges=req_doc_ranges,
             for_cudagraph_capture=for_capture,
             rswa_prefix_lens=input_batch.prompt_lens,
+            num_computed_tokens_cpu=torch.from_numpy(num_computed_tokens_cpu),
```
[the remark this answers is not in the document]
Comment (ehartford) [reply vllm/v1/worker/gpu/model_states/default.py]: Addressed in 35892949d: prepare_attn now stages num_computed_tokens_cpu (and is_prefilling) through persistent padded buffers — no per-step allocation — and the steady-cache hit path explicitly refreshes cm._num_computed_tokens_cpu alongside max_seq_len, so a cache hit can never replay first-step computed-token counts even if the producer's buffer-view pattern changes.

File: vllm/models/deepseek_v4/metal_indexer.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+
+        tile_k = min(k_eff, k1 - k0)
+        vals, idx = _topk_desc_stable(logits, tile_k)
+        idx += k0
+        # Retained survivors precede the new tile positionally and always
+        # carry smaller global indices (tiles ascend), so the stable merge
+        # preserves (logit desc, index asc) globally by induction.
+        merged_vals = torch.cat((best_vals, vals), dim=1)
+        merged_idx = torch.cat((best_idx, idx), dim=1)
+        best_vals, keep = _topk_desc_stable(merged_vals, k_eff)
```
[the remark this answers is not in the document]
Comment (ehartford) [reply vllm/models/deepseek_v4/metal_indexer.py:168]: Addressed in 35892949d: _score_and_select_streaming now slices the candidate axis for both K layouts ([n_k, 128] and [rows, n_k, 128] with per-row scales) and additionally accepts a k_provider tile fetch, which the decode eager fallback uses to keep its cache gather bounded to 16 MiB tiles. Parity tests cover the per-row layout pre-gathered and via the provider, tie plateaus included.
